File: Sources/DemoTape/DemoControl.swift
import Foundation
import CoreGraphics

/// External control surface for DemoTape, so an orchestrator (e.g. Kiro driving a browser with
/// Playwright, or a computer-use agent) can run a demo hands-off: **start recording → drive the
/// app → stop → collect the finished video** — without embedding any of that logic (or any
/// third-party dependency) inside DemoTape itself.
///
/// Control comes in over a `demotape://` URL (handled in `AppDelegate`); progress goes back out
/// via a small pollable `control.json` status file. This file holds the **pure, testable** URL
/// parsing; the side effects (starting capture, writing status) live in `AppDelegate`.
///
/// URL grammar:
///   demotape://record/start                         full screen, 3-2-1 countdown
///   demotape://record/start?countdown=0             start immediately (best for automation)
///   demotape://record/start?mode=area&x=&y=&w=&h=   crop to a pixel rect on the main display
///   demotape://record/start?nx=&ny=&nw=&nh=         crop to a normalized rect (0…1, top-left)
///   demotape://record/start?mic=1&webcam=0          override input toggles for this take
///   demotape://record/stop                          stop + auto-render
enum DemoControl {

    /// Where to crop the capture.
    enum Region: Equatable {
        case fullScreen
        case normalized(CGRect)   // 0…1, top-left origin
        case pixels(CGRect)       // device points, top-left origin, main display
    }

    struct StartOptions: Equatable {
        var region: Region = .fullScreen
        var countdown: Int = 3    // seconds; 0 = begin immediately
        var microphone: Bool? = nil   // nil = leave the current setting
        var webcam: Bool? = nil
    }

    enum Command: Equatable {
        case start(StartOptions)
        case stop
        /// Move (and optionally click) the cursor from the RUNNING app process, which holds the
        /// Accessibility grant — so synthetic clicks register and trigger auto-zoom. `click` false
        /// = move only.
        /// `glideMs` > 0 animates the trip instead of teleporting: eased, slightly arced, with a
        /// small overshoot-and-settle at the end. A hard warp reads as a robot on video, and the
        /// travel itself is what carries the viewer's eye between two points, so a demo should
        /// almost always glide. The animation runs inside the app (one URL per move) because
        /// issuing a URL per intermediate point would stutter at ~150ms of shell overhead each.
        case cursor(x: Double, y: Double, click: Bool, glideMs: Int = 0)
        /// Type text as REAL OS keystrokes, from the running app (which holds the Accessibility
        /// grant). This matters for more than fidelity: auto-zoom is driven by clicks *and keys*, so
        /// keystrokes the app can't observe leave the camera wide while the user types. A browser
        /// automation tool's synthetic typing is invisible to our global event monitor, so the text
        /// appeared with no zoom held on the field. Typing through the app records real KeySamples,
        /// which hold the zoom exactly as they do when a human types.
        /// `cps` is characters per second (0 = a sensible human default).
        ///
        /// `expectedApp` is a REQUIRED-in-practice safety catch: synthetic keystrokes go to whichever
        /// window holds system focus, not to any window the caller has a handle on. If focus has
        /// moved, the text lands in the user's editor or terminal — which can be destructive, and is
        /// invisible to the caller because the keys still get recorded. When set, the app refuses to
        /// type unless that application is frontmost.
        case type(text: String, cps: Double, expectedApp: String?)
        /// Record typing activity WITHOUT posting keystrokes, so the auto-zoom holds on the focused
        /// field while some other tool types the visible text.
        ///
        /// This exists because browsers reject synthetic key events that carry no virtual keycode:
        /// posting real keystrokes at a Chromium window types nothing, and if another window is
        /// frontmost it types there instead. So a browser-driven demo must let the automation tool
        /// enter the text — and then nothing tells DemoTape that typing is happening, leaving the
        /// camera to drift off the field mid-sentence. `chars` is how many characters are being
        /// typed and `cps` the rate, which together reproduce the same zoom hold real typing gives.
        case typingActivity(chars: Int, cps: Double)
        /// Open one of DemoTape's own windows. This exists so a walkthrough *of DemoTape itself*
        /// can be scripted: an orchestrator can't reliably click menu rows (their screen rects
        /// aren't discoverable from outside), but it can ask the app to show a window directly.
        /// `holdMs` only applies to `.menu`: an open NSMenu runs a nested tracking loop that stalls
        /// incoming control URLs until it closes, so the app must dismiss its own menu after the
        /// requested time. Without it a scripted walkthrough deadlocks behind its own menu.
        case openUI(Window, holdMs: Int = 0)
        /// Resolve a target from the LIVE UI (by label/role, via AppKit for our own windows or the
        /// Accessibility API for any other app) and act on it. This is the coordinate-free path:
        /// `find` publishes the element's rect, `click` glides to it and clicks it.
        case element(query: UIQuery.Query, click: Bool)
        /// Write the current UI tree to `ui-tree.json` so an agent can read what's on screen.
        case dumpUI(app: String?)
    }

    /// Windows reachable through `demotape://ui/open?window=…`.
    ///
    /// `menu` isn't a window — it drops the menu-bar menu open. Clicking the icon with a synthetic
    /// event doesn't reliably reach a status item, so the app performs the click on itself instead,
    /// which needs no screen coordinates at all.
    enum Window: String, Equatable, CaseIterable {
        case about, publish, composer, settings, welcome, menu
    }

    /// Parses a `demotape://` control URL into a command. Returns nil for anything unrecognized.
    static func parse(_ url: URL) -> Command? {
        guard url.scheme?.lowercased() == "demotape" else { return nil }
        guard let comps = URLComponents(url: url, resolvingAgainstBaseURL: false) else { return nil }

        // Action can be the host (demotape://stop) or a path segment (demotape://record/stop).
        var tokens: [String] = []
        if let host = comps.host { tokens.append(host.lowercased()) }
        tokens += comps.path.split(separator: "/").map { $0.lowercased() }

        // Query lookup (case-insensitive keys). `+` is decoded as a space: URLComponents leaves it
        // literal, but callers building URLs from shells and scripts routinely use it for spaces,
        // which otherwise silently fails to match a label like "Check for Updates".
        var q: [String: String] = [:]
        for item in comps.queryItems ?? [] {
            q[item.name.lowercased()] = item.value?.replacingOccurrences(of: "+", with: " ")
        }
        func dbl(_ k: String) -> Double? { q[k].flatMap(Double.init) }

        if tokens.contains("typing") {
            // demotape://typing?chars=42&cps=14  — activity only, nothing is posted.
            guard let chars = dbl("chars"), chars > 0 else { return nil }
            return .typingActivity(chars: Int(chars), cps: max(0, dbl("cps") ?? 0))
        }
        if tokens.contains("type") {
            // demotape://type?text=hello%20world&cps=14&app=Chromium
            guard let text = q["text"], !text.isEmpty else { return nil }
            return .type(text: text, cps: max(0, dbl("cps") ?? 0), expectedApp: q["app"])
        }
        if tokens.contains("cursor") {
            guard let x = dbl("x"), let y = dbl("y") else { return nil }
            // demotape://cursor/glide?x=&y=&ms=420  (ms also accepted as `glide`)
            var glide = Int(dbl("ms") ?? dbl("glide") ?? 0)
            if tokens.contains("glide"), glide <= 0 { glide = 420 }   // sensible default travel
            return .cursor(x: x, y: y, click: tokens.contains("click"), glideMs: max(0, glide))
        }
        if tokens.contains("ui") {
            // demotape://ui/dump?app=Safari
            if tokens.contains("dump") { return .dumpUI(app: q["app"]) }
            // Semantic targeting: demotape://ui/click?label=Export&role=AXButton&app=Safari
            // (or ui/find to only publish the rect). Beats coordinates: it survives a moved window
            // and reports honestly when the element isn't there.
            if tokens.contains("click") || tokens.contains("find") {
                guard let label = q["label"], !label.isEmpty else { return nil }
                let query = UIQuery.Query(label: label,
                                          role: q["role"],
                                          app: q["app"],
                                          index: Int(q["index"] ?? "") ?? 0)
                return .element(query: query, click: tokens.contains("click"))
            }
            // demotape://ui/open?window=about  — or the shorthand demotape://ui/about
            let named = q["window"]?.lowercased() ?? tokens.last
            guard let named = named, let win = Window(rawValue: named) else { return nil }
            let hold = Int(dbl("hold") ?? dbl("holdms") ?? 0)
            return .openUI(win, holdMs: max(0, hold))
        }
        if tokens.contains("stop") { return .stop }
        guard tokens.contains("start") else { return nil }
        func flag(_ keys: [String]) -> Bool? {
            for k in keys { if let v = q[k]?.lowercased() {
                if ["1", "true", "yes", "on"].contains(v) { return true }
                if ["0", "false", "no", "off"].contains(v) { return false }
            } }
            return nil
        }

        var opts = StartOptions()
        if let nx = dbl("nx"), let ny = dbl("ny"), let nw = dbl("nw"), let nh = dbl("nh") {
            opts.region = .normalized(CGRect(x: nx, y: ny, width: nw, height: nh))
        } else if let x = dbl("x"), let y = dbl("y"), let w = dbl("w"), let h = dbl("h") {
            opts.region = .pixels(CGRect(x: x, y: y, width: w, height: h))
        } else {
            opts.region = .fullScreen   // also the case for mode=fullscreen
        }
        if let c = q["countdown"], let n = Int(c) { opts.countdown = max(0, n) }
        opts.microphone = flag(["mic", "microphone"])
        opts.webcam = flag(["webcam", "cam", "camera"])
        return .start(opts)
    }

    // MARK: - Status file (pollable by the orchestrator)

    /// Path of the status file the orchestrator polls (`DemoTape/.demotape/control.json`).
    static var statusURL: URL { Paths.supportDirectory.appendingPathComponent("control.json") }

    /// Writes the current control state. `state` is one of idle/countdown/recording/rendering;
    /// `lastOutput` (when known) is the absolute path of the most recent finished video.
    /// `menuBarButton` (when known) is the screen rect of DemoTape's menu-bar icon in top-left
    /// origin points, so an orchestrator can click it to film the menu — there's no other way to
    /// find a menu-bar extra's position from outside the app.
    /// `window` (when known) is the screen rect of the window most recently opened through
    /// `ui/open`, in the same top-left origin points the cursor commands take. A scripted
    /// walkthrough needs this to aim: windows are centred by AppKit, so guessing their position
    /// risks landing an "emphasis" click on a real button.
    /// Result of the most recent semantic lookup, so a script can confirm what it hit — or that it
    /// missed, which it must treat as a failure rather than clicking blindly.
    struct ElementStatus {
        var role: String
        var label: String
        var frame: CGRect
        var found: Bool
    }

    static func writeStatus(state: String, lastOutput: String? = nil,
                            menuBarButton: CGRect? = nil, window: CGRect? = nil,
                            element: ElementStatus? = nil) {
        var dict: [String: Any] = [
            "state": state,
            "recording": (state == "recording"),
            "updatedAt": ISO8601DateFormatter().string(from: Date())
        ]
        if let lastOutput = lastOutput { dict["lastOutput"] = lastOutput }
        if let r = menuBarButton {
            dict["menuBarButton"] = ["x": r.minX, "y": r.minY, "w": r.width, "h": r.height,
                                     "centerX": r.midX, "centerY": r.midY]
        }
        if let r = window {
            dict["window"] = ["x": r.minX, "y": r.minY, "w": r.width, "h": r.height,
                              "centerX": r.midX, "centerY": r.midY]
        }
        if let e = element {
            dict["element"] = ["found": e.found, "role": e.role, "label": e.label,
                               "x": e.frame.minX, "y": e.frame.minY,
                               "w": e.frame.width, "h": e.frame.height,
                               "centerX": e.frame.midX, "centerY": e.frame.midY]
        }
        guard let data = try? JSONSerialization.data(withJSONObject: dict, options: [.prettyPrinted]) else { return }
        try? data.write(to: statusURL, options: .atomic)
    }
}

File: windows/src/App/Domain/Control/DemoControl.cs
using System.Globalization;

namespace DemoTape.Domain.Control;

/// <summary>
/// External control surface for DemoTape, ported from the macOS <c>DemoControl.swift</c>. It lets an
/// orchestrator (e.g. Kiro driving a browser with Playwright, or a computer-use agent) run a demo
/// hands-off: <b>start recording → drive the app → stop → collect the finished video</b> — without
/// embedding any of that logic (or any third-party dependency) inside DemoTape itself.
///
/// <para>Control comes in over a <c>demotape://</c> URL (registered + dispatched by the app layer);
/// progress goes back out via a small pollable <c>control.json</c> status file. This class holds the
/// <b>pure, testable</b> URL parsing; the side effects (starting capture, writing status) live in the
/// app/infrastructure layer.</para>
///
/// <para>URL grammar (parity with macOS):</para>
/// <code>
///   demotape://record/start                         full screen, 3-2-1 countdown
///   demotape://record/start?countdown=0             start immediately (best for automation)
///   demotape://record/start?mode=area&amp;x=&amp;y=&amp;w=&amp;h= crop to a pixel rect on the main display
///   demotape://record/start?nx=&amp;ny=&amp;nw=&amp;nh=          crop to a normalized rect (0…1, top-left)
///   demotape://record/start?mic=1&amp;webcam=0          override input toggles for this take
///   demotape://record/webcam                        talking-head to camera, no screen
///   demotape://record/stop                          stop + auto-render
/// </code>
/// </summary>
public static class DemoControl
{
    /// <summary>Where to crop the capture.</summary>
    public enum RegionKind { FullScreen, Normalized, Pixels }

    /// <summary>A capture region: full screen, or a rectangle in normalized (0…1) or pixel units.</summary>
    public readonly record struct Region(RegionKind Kind, double X, double Y, double W, double H)
    {
        public static Region FullScreen => new(RegionKind.FullScreen, 0, 0, 0, 0);
        public static Region Normalized(double x, double y, double w, double h) => new(RegionKind.Normalized, x, y, w, h);
        public static Region Pixels(double x, double y, double w, double h) => new(RegionKind.Pixels, x, y, w, h);
    }

    /// <summary>Options for a <c>record/start</c> command. Null mic/webcam = leave the current setting.</summary>
    public sealed record StartOptions
    {
        public Region Region { get; init; } = Region.FullScreen;
        public int Countdown { get; init; } = 3;   // seconds; 0 = begin immediately
        public bool? Microphone { get; init; }
        public bool? Webcam { get; init; }
        /// <summary>Record the camera only (talking-head), not the screen. <c>demotape://record/webcam</c>.</summary>
        public bool WebcamOnly { get; init; }
    }

    /// <summary>The kind of a parsed control command.</summary>
    public enum CommandKind { Start, Stop, Cursor, Type, TypingActivity, CursorPath, OpenUi, Element, DumpUi }

    /// <summary>Windows reachable through <c>demotape://ui/open?window=…</c>.</summary>
    public enum UiWindow { About, Publish, Composer, Settings, Welcome, Voiceover, Captions, Menu }

    /// <summary>A single point (screen points, top-left origin).</summary>
    public readonly record struct Point(double X, double Y);

    /// <summary>A semantic UI target resolved from the live UI (label/role/app/index).</summary>
    public sealed record ElementQuery(string Label, string? Role = null, string? App = null, int Index = 0);

    /// <summary>
    /// A parsed <c>demotape://</c> command. A discriminated union in Swift; here a single record whose
    /// <see cref="Kind"/> selects which fields are meaningful. Use the static factories to build.
    /// </summary>
    public sealed record Command
    {
        public CommandKind Kind { get; private init; }
        public StartOptions? Start { get; private init; }
        public double CursorX { get; private init; }
        public double CursorY { get; private init; }
        public bool Click { get; private init; }
        public int GlideMs { get; private init; }
        public string? Text { get; private init; }
        public double Cps { get; private init; }
        public string? ExpectedApp { get; private init; }
        public int Chars { get; private init; }
        public Point? Caret { get; private init; }
        public IReadOnlyList<Point> PathPoints { get; private init; } = Array.Empty<Point>();
        public int PathMs { get; private init; }
        public UiWindow Window { get; private init; }
        public int HoldMs { get; private init; }
        public ElementQuery? Query { get; private init; }
        public string? App { get; private init; }

        internal static Command MakeStart(StartOptions o) => new() { Kind = CommandKind.Start, Start = o };
        internal static Command MakeStop() => new() { Kind = CommandKind.Stop };
        internal static Command MakeCursor(double x, double y, bool click, int glideMs) =>
            new() { Kind = CommandKind.Cursor, CursorX = x, CursorY = y, Click = click, GlideMs = glideMs };
        internal static Command MakeType(string text, double cps, string? app) =>
            new() { Kind = CommandKind.Type, Text = text, Cps = cps, ExpectedApp = app };
        internal static Command MakeTypingActivity(int chars, double cps, Point? caret) =>
            new() { Kind = CommandKind.TypingActivity, Chars = chars, Cps = cps, Caret = caret };
        internal static Command MakeCursorPath(IReadOnlyList<Point> points, int ms) =>
            new() { Kind = CommandKind.CursorPath, PathPoints = points, PathMs = ms };
        internal static Command MakeOpenUi(UiWindow window, int holdMs) =>
            new() { Kind = CommandKind.OpenUi, Window = window, HoldMs = holdMs };
        internal static Command MakeElement(ElementQuery q, bool click) =>
            new() { Kind = CommandKind.Element, Query = q, Click = click };
        internal static Command MakeDumpUi(string? app) => new() { Kind = CommandKind.DumpUi, App = app };
    }

    /// <summary>
    /// Parses a <c>demotape://</c> control URL into a <see cref="Command"/>. Returns null for anything
    /// unrecognized or for a foreign scheme.
    /// </summary>
    public static Command? Parse(Uri? url)
    {
        if (url is null) return null;
        if (!string.Equals(url.Scheme, "demotape", StringComparison.OrdinalIgnoreCase)) return null;

        // Action can be the host (demotape://stop) or a path segment (demotape://record/stop).
        var tokens = new List<string>();
        if (!string.IsNullOrEmpty(url.Host)) tokens.Add(url.Host.ToLowerInvariant());
        foreach (var seg in url.AbsolutePath.Split('/', StringSplitOptions.RemoveEmptyEntries))
            tokens.Add(seg.ToLowerInvariant());

        // Query lookup (case-insensitive keys). Parsed by hand to stay dependency-free. Percent
        // escapes are decoded, then `+` decodes as a space — shells and scripts routinely use `+`
        // for spaces, which would otherwise fail to match a label like "Check for Updates".
        var q = new Dictionary<string, string>(StringComparer.OrdinalIgnoreCase);
        var query = url.Query.TrimStart('?');
        foreach (var pair in query.Split('&', StringSplitOptions.RemoveEmptyEntries))
        {
            var eq = pair.IndexOf('=');
            var rawKey = eq >= 0 ? pair[..eq] : pair;
            var rawVal = eq >= 0 ? pair[(eq + 1)..] : string.Empty;
            var key = Uri.UnescapeDataString(rawKey).Replace('+', ' ');
            var val = Uri.UnescapeDataString(rawVal).Replace('+', ' ');
            q[key] = val;
        }

        double? Dbl(string k) =>
            q.TryGetValue(k, out var v) && double.TryParse(v, NumberStyles.Float, CultureInfo.InvariantCulture, out var d)
                ? d : (double?)null;

        bool Has(string t) => tokens.Contains(t);

        if (Has("typing"))
        {
            // demotape://typing?chars=42&cps=14[&x=&y=]  — activity only, nothing is posted.
            var chars = Dbl("chars");
            if (chars is null || chars <= 0) return null;
            Point? caret = null;
            var cx = Dbl("x"); var cy = Dbl("y");
            if (cx is not null && cy is not null) caret = new Point(cx.Value, cy.Value);
            return Command.MakeTypingActivity((int)chars.Value, Math.Max(0, Dbl("cps") ?? 0), caret);
        }
        if (Has("type"))
        {
            // demotape://type?text=hello%20world&cps=14&app=Chromium
            if (!q.TryGetValue("text", out var text) || string.IsNullOrEmpty(text)) return null;
            q.TryGetValue("app", out var app);
            return Command.MakeType(text, Math.Max(0, Dbl("cps") ?? 0), string.IsNullOrEmpty(app) ? null : app);
        }
        if (Has("path"))
        {
            // demotape://cursor/path?pts=120,80;200,140;260,90&ms=1400
            if (!q.TryGetValue("pts", out var raw) || string.IsNullOrEmpty(raw)) return null;
            var points = new List<Point>();
            foreach (var pair in raw.Split(';', StringSplitOptions.RemoveEmptyEntries))
            {
                var parts = pair.Split(',');
                if (parts.Length != 2) continue;
                if (double.TryParse(parts[0].Trim(), NumberStyles.Float, CultureInfo.InvariantCulture, out var px) &&
                    double.TryParse(parts[1].Trim(), NumberStyles.Float, CultureInfo.InvariantCulture, out var py))
                    points.Add(new Point(px, py));
            }
            if (points.Count < 2) return null;
            int ms = (int)(Dbl("ms") ?? 0);
            return Command.MakeCursorPath(points, Math.Max(120, ms == 0 ? 900 : ms));
        }
        if (Has("cursor"))
        {
            var x = Dbl("x"); var y = Dbl("y");
            if (x is null || y is null) return null;
            // demotape://cursor/glide?x=&y=&ms=420  (ms also accepted as `glide`)
            int glide = (int)(Dbl("ms") ?? Dbl("glide") ?? 0);
            if (Has("glide") && glide <= 0) glide = 420;   // sensible default travel
            return Command.MakeCursor(x.Value, y.Value, Has("click"), Math.Max(0, glide));
        }
        if (Has("ui"))
        {
            // demotape://ui/dump?app=Safari
            if (Has("dump")) { q.TryGetValue("app", out var dumpApp); return Command.MakeDumpUi(string.IsNullOrEmpty(dumpApp) ? null : dumpApp); }
            // Semantic targeting: demotape://ui/click?label=Export&role=AXButton&app=Safari
            if (Has("click") || Has("find"))
            {
                if (!q.TryGetValue("label", out var label) || string.IsNullOrEmpty(label)) return null;
                q.TryGetValue("role", out var role);
                q.TryGetValue("app", out var app);
                int index = q.TryGetValue("index", out var idx) && int.TryParse(idx, out var i) ? i : 0;
                var elementQuery = new ElementQuery(label,
                    string.IsNullOrEmpty(role) ? null : role,
                    string.IsNullOrEmpty(app) ? null : app,
                    index);
                return Command.MakeElement(elementQuery, Has("click"));
            }
            // demotape://ui/open?window=about  — or the shorthand demotape://ui/about
            string? named = q.TryGetValue("window", out var w) && !string.IsNullOrEmpty(w)
                ? w.ToLowerInvariant()
                : (tokens.Count > 0 ? tokens[^1] : null);
            if (named is null || !TryParseWindow(named, out var win)) return null;
            int hold = (int)(Dbl("hold") ?? Dbl("holdms") ?? 0);
            return Command.MakeOpenUi(win, Math.Max(0, hold));
        }
        if (Has("stop")) return Command.MakeStop();

        // demotape://record/webcam[?countdown=&mic=]  — talking-head to camera, no screen.
        bool webcamOnly = Has("webcam");
        if (!Has("start") && !webcamOnly) return null;

        bool? Flag(params string[] keys)
        {
            foreach (var k in keys)
            {
                if (q.TryGetValue(k, out var v))
                {
                    var lv = v.ToLowerInvariant();
                    if (lv is "1" or "true" or "yes" or "on") return true;
                    if (lv is "0" or "false" or "no" or "off") return false;
                }
            }
            return null;
        }

        Region region;
        var nx = Dbl("nx"); var ny = Dbl("ny"); var nw = Dbl("nw"); var nh = Dbl("nh");
        if (nx is not null && ny is not null && nw is not null && nh is not null)
            region = Region.Normalized(nx.Value, ny.Value, nw.Value, nh.Value);
        else
        {
            var px = Dbl("x"); var py = Dbl("y"); var pw = Dbl("w"); var ph = Dbl("h");
            if (px is not null && py is not null && pw is not null && ph is not null)
                region = Region.Pixels(px.Value, py.Value, pw.Value, ph.Value);
            else
                region = Region.FullScreen;   // also the case for mode=fullscreen
        }

        int countdown = 3;
        if (q.TryGetValue("countdown", out var cv) && int.TryParse(cv, out var cn)) countdown = Math.Max(0, cn);

        var opts = new StartOptions
        {
            Region = region,
            Countdown = countdown,
            Microphone = Flag("mic", "microphone"),
            Webcam = webcamOnly ? null : Flag("webcam", "cam", "camera"),
            WebcamOnly = webcamOnly,
        };
        return Command.MakeStart(opts);
    }

    private static bool TryParseWindow(string name, out UiWindow window)
    {
        switch (name.ToLowerInvariant())
        {
            case "about": window = UiWindow.About; return true;
            case "publish": window = UiWindow.Publish; return true;
            case "composer": window = UiWindow.Composer; return true;
            case "settings": window = UiWindow.Settings; return true;
            case "welcome": window = UiWindow.Welcome; return true;
            case "voiceover": window = UiWindow.Voiceover; return true;
            case "captions": window = UiWindow.Captions; return true;
            case "menu": window = UiWindow.Menu; return true;
            default: window = default; return false;
        }
    }
}
